File: src/main/java/com/blasphemy/config/ModConfig.java
package com.blasphemy.config;

import com.blasphemy.Blasphemy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import net.fabricmc.loader.api.FabricLoader;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;

/**
 * 模组配置类
 * 负责加载和保存模组配置
 */
public class ModConfig {
    
    private static ModConfig instance;
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private static final File CONFIG_FILE = FabricLoader.getInstance().getConfigDir().resolve("blasphemy.json").toFile();
    
    // 炽天使配置
    public SeraphimSwordConfig seraphimSword = new SeraphimSwordConfig();
    
    // 激流之剑配置
    public RapidsSwordConfig rapidsSword = new RapidsSwordConfig();
    
    // 妖刀村正配置
    public MuramasaSwordConfig muramasaSword = new MuramasaSwordConfig();
    
    // 群体斩击附魔配置
    public CleaveConfig cleaveConfig = new CleaveConfig();
    
    // 传送门配置
    public PortalConfig portalConfig = new PortalConfig();
    
    /**
     * 加载配置
     */
    public static void load() {
        Blasphemy.LOGGER.info("加载配置文件...");
        
        // 如果配置文件不存在，创建默认配置
        if (!CONFIG_FILE.exists()) {
            instance = new ModConfig();
            save();
            Blasphemy.LOGGER.info("创建默认配置文件");
            return;
        }
        
        // 读取配置文件
        try (FileReader reader = new FileReader(CONFIG_FILE)) {
            instance = GSON.fromJson(reader, ModConfig.class);
            Blasphemy.LOGGER.info("成功加载配置文件");
        } catch (IOException e) {
            Blasphemy.LOGGER.error("加载配置文件失败", e);
            instance = new ModConfig();
        }
    }
    
    /**
     * 保存配置
     */
    public static void save() {
        try (FileWriter writer = new FileWriter(CONFIG_FILE)) {
            GSON.toJson(instance, writer);
            Blasphemy.LOGGER.info("成功保存配置文件");
        } catch (IOException e) {
            Blasphemy.LOGGER.error("保存配置文件失败", e);
        }
    }
    
    /**
     * 获取配置实例
     */
    public static ModConfig getConfig() {
        if (instance == null) {
            load();
        }
        return instance;
    }
    
    /**
     * 炽天使配置类
     */
    public static class SeraphimSwordConfig {
        public float baseDamage = 7.0f;
        public float attackSpeed = -2.4f;
        public float specialDamageMultiplier = 0.1f;
        public float pushStrength = 1.5f;
        public int cooldownSeconds = 60;
        public int particleCount = 40;
        public boolean enableParticles = true;
        public float undeadDamageMultiplier = 0.3f;
        public List<String> undeadEntityTags = Arrays.asList("minecraft:undead", "forge:undead", "c:undead");
    }
    
    /**
     * 激流之剑配置类
     */
    public static class RapidsSwordConfig {
        public float baseDamage = 6.0f;
        public float attackSpeed = -1.9f;
        public float specialDamageMultiplier = 0.05f;
        public float executionThreshold = 0.2f;
        public float executionChance = 0.3f;
    }
    
    /**
     * 妖刀村正配置类
     */
    public static class MuramasaSwordConfig {
        public float baseDamage = 8.0f;
        public float attackSpeed = -2.6f;
        public float specialDamageMultiplier = 0.025f;
        public float executionThreshold = 0.2f;
        public float executionChance = 0.3f;
        public float armorThreshold = 10.0f;
    }
    
    /**
     * 群体斩击附魔配置类
     */
    public static class CleaveConfig {
        public float attackRange = 3.0f;
        public float baseMultiplier = 0.15f;
        public float levelMultiplier = 0.05f;
        public int maxLevel = 5;
    }
    
    /**
     * 传送门配置类
     */
    public static class PortalConfig {
        public boolean enabled = true;
        public boolean supportVanillaItems = true;
        public List<String> portalBlocks = Arrays.asList("minecraft:crying_obsidian", "minecraft:diamond_block");
        public String ignitionItem = "minecraft:heart_of_the_sea";
        public PortalMessages messages = new PortalMessages();
        
        public static class PortalMessages {
            public String invalidBlock = "message.blasphemy.portal.invalid_block";
            public String invalidItem = "message.blasphemy.portal.invalid_item";
        }
    }
}

File: src/main/kotlin/com/blasphemy/config/ModConfig.kt
package com.blasphemy.config

import com.blasphemy.Blasphemy
import com.google.gson.GsonBuilder
import java.io.FileReader
import java.io.FileWriter
import net.fabricmc.loader.api.FabricLoader

/** 模组配置类，负责管理和存储模组配置 */
object ModConfig {

    // 配置文件
    private val configFile =
            FabricLoader.getInstance().configDir.resolve("${Blasphemy.MOD_ID}.json").toFile()

    // 默认配置
    private val defaultConfig = Config()

    // 当前配置
    private var config = defaultConfig.copy()

    /** 加载配置 */
    fun load() {
        if (configFile.exists()) {
            try {
                FileReader(configFile).use { reader ->
                    val gson = GsonBuilder().setPrettyPrinting().create()
                    config = gson.fromJson(reader, Config::class.java)
                    Blasphemy.logger.info("配置文件已加载")
                }
            } catch (e: Exception) {
                Blasphemy.logger.error("加载配置文件失败，使用默认配置", e)
                config = defaultConfig.copy()
                save() // 保存默认配置
            }
        } else {
            Blasphemy.logger.info("配置文件不存在，创建默认配置")
            save() // 保存默认配置
        }
    }

    /** 保存配置 */
    fun save() {
        try {
            if (!configFile.parentFile.exists()) {
                configFile.parentFile.mkdirs()
            }

            FileWriter(configFile).use { writer ->
                val gson = GsonBuilder().setPrettyPrinting().create()
                gson.toJson(config, writer)
            }
            Blasphemy.logger.info("配置文件已保存")
        } catch (e: Exception) {
            Blasphemy.logger.error("保存配置文件失败", e)
        }
    }

    /** 获取当前配置 */
    fun getConfig(): Config {
        return config
    }

    /** 配置数据类 */
    data class Config(
            // 天使剑配置
            val seraphimSword: SeraphimSwordConfig = SeraphimSwordConfig(),

            // 急流剑配置
            val rapidsSword: RapidsSwordConfig = RapidsSwordConfig(),

            // 村正剑配置
            val muramasaSword: MuramasaSwordConfig = MuramasaSwordConfig(),

            // 群体斩击附魔配置
            val cleaveConfig: CleaveConfig = CleaveConfig(),

            // 传送门配置
            val portalConfig: PortalConfig = PortalConfig()
    )

    /** 天使剑配置数据类 */
    data class SeraphimSwordConfig(
            val baseDamage: Int = 7, // 基础伤害
            val attackSpeed: Float = -2.4f, // 攻击速度
            val specialDamageMultiplier: Float = 0.1f, // 特殊伤害乘数
            val pushStrength: Float = 1.5f, // 推力强度
            val cooldownSeconds: Int = 60, // 冷却时间
            val particleCount: Int = 40, // 粒子效果数量
            val enableParticles: Boolean = true, // 是否启用粒子效果
            val undeadDamageMultiplier: Float = 0.3f, // 亡灵伤害倍率
            val undeadEntityTags: List<String> =
                    listOf( // 亡灵实体标签列表
                            "minecraft:undead",
                            "forge:undead",
                            "c:undead"
                    )
    )

    /** 急流剑配置数据类 */
    data class RapidsSwordConfig(
            val baseDamage: Int = 6, // 基础伤害
            val attackSpeed: Float = -1.9f, // 攻击速度
            val specialDamageMultiplier: Float = 0.05f, // 特殊伤害乘数
            val executionThreshold: Float = 0.2f, // 斩杀阈值
            val executionChance: Float = 0.3f // 斩杀几率
    )

    /** 村正剑配置数据类 */
    data class MuramasaSwordConfig(
            val baseDamage: Int = 8, // 基础伤害
            val attackSpeed: Float = -2.6f, // 攻击速度
            val specialDamageMultiplier: Float = 0.025f, // 特殊伤害乘数
            val executionThreshold: Float = 0.2f, // 斩杀阈值
            val executionChance: Float = 0.3f, // 斩杀几率
            val armorThreshold: Int = 10 // 护甲阈值
    )

    /** 群体斩击附魔配置数据类 */
    data class CleaveConfig(
            val attackRange: Double = 3.0, // 攻击范围（方块）
            val baseMultiplier: Float = 0.15f, // 基础伤害倍率（15%）
            val levelMultiplier: Float = 0.05f, // 每级增加的伤害倍率（5%）
            val maxLevel: Int = 5 // 最大等级
    )

    /** 传送门配置数据类 */
    data class PortalConfig(
            val enabled: Boolean = true, // 是否启用自定义传送门
            val portalBlocks: List<String> =
                    listOf( // 允许的传送门方块ID列表
                            "minecraft:obsidian",
                            "minecraft:crying_obsidian"
                    ),
            val ignitionItem: String = "minecraft:flint_and_steel", // 开启传送门的物品ID
            val messages: PortalMessages = PortalMessages() // 提示语配置
    )

    /** 传送门提示语配置数据类 */
    data class PortalMessages(
            val invalidBlock: String = "message.blasphemy.portal.invalid_block", // 使用错误方块时的提示
            val invalidItem: String = "message.blasphemy.portal.invalid_item", // 使用错误物品时的提示
            val portalCreated: String = "message.blasphemy.portal.created" // 传送门创建成功时的提示
    )
}
